[BE] Update ProcessGroupWrapper to add deserializer and improve logs

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#79724 [BE] Update ProcessGroupWrapper to add deserializer and improve logs**
* #79723 [small changes] Add closing parentheses to CollectiveFingerprint output

Fixes #78022
And solves #72435 improvement #3

### Current Error Message:
`Detected mismatch between collectives on ranks. Rank 3 is running inconsistent collective: CollectiveFingerPrint(OpType=BARRIER)`

### New Error Message (shows rank and collective):
`Detected mismatch between collectives on ranks. Rank 0 is running collective: CollectiveFingerPrint(OpType=ALLREDUCE, …), but Rank 1 is running collective: REDUCE.`

### Changes:
- Add a deserialize function to deserialize a tensor to a CollectiveFingerprint
- Update log message to output the RANK and COLLECTIVE TYPE

### Context: 
ProcessGroupWrapper is used when TORCH_DISTRIBUTED_DEBUG=DETAIL and used to detect mismatched collectives. It performs this by serializing each collective call (Operation type, tensor data types, device types, shapes) into a tensor, performing an AllGather among all other process group wrappers, then detecting whether any of these tensors are mismatched.


**TODO**: The log message can also log the tensor data type, shapes, devices as well using data from deserialize method. Will file an issue

H-Huang committed:
[BE] Update ProcessGroupWrapper to add deserializer and improve logs

[ghstack-poisoned]

H-Huang commented on torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -62,18 +72,56 @@ struct CollectiveFingerPrint {
+  // Takes a serialized fingerprint from
+  // CollectiveFingerPrint::serialize_fingerprint and deserializes it back to a
+  // CollectiveFingerPrint struct
+  CollectiveFingerPrint deserialize_fingerprint(at::Tensor serialized_tensor) {
```
Thanks Rodrigo, I agree with you the manual deserialization is not ideal.

I think the challenge of using a library like flatbuffers / protobuf is that it also requires a change in how we want the nodes to communicate. Since we right now using `allgather` to transfer information between process group wrappers, then the data needs to be in a tensor form. If we use a 3rd party library we would either need to also to somehow transform the serialized data into tensor to continue using `allgather` or communicate the serialized data separately somehow.

rohan-varma reviewed (approved):
Looks great, thanks for putting this together!

Can we have an example of what the new error msg looks like?

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -62,18 +72,56 @@ struct CollectiveFingerPrint {
+    // deserializing
+    int index = 0;
+    // 1. OpType
+    OpType optype = OpType(serialized_tensor[index].item<int>());
```
might want to add some validation to ensure we don't access out of bounds and have a hard to debug error.

H-Huang commented:
> Can we have an example of what the new error msg looks like?

Yep I wrote it in the description, but I will move it to the top of the text so it is clearer

H-Huang committed:
Update on "[BE] Update ProcessGroupWrapper to add deserializer and improve logs"


Fixes #78022
And solves #72435 improvement #3

### Current Error Message:
`Detected mismatch between collectives on ranks. Rank 3 is running inconsistent collective: CollectiveFingerPrint(OpType=BARRIER)`

### New Error Message (shows rank and collective):
`Detected mismatch between collectives on ranks. Rank 0 is running collective: CollectiveFingerPrint(OpType=ALLREDUCE, …), but Rank 1 is running collective: REDUCE.`

### Changes:
- Add a deserialize function to deserialize a tensor to a CollectiveFingerprint
- Update log message to output the RANK and COLLECTIVE TYPE

### Context: 
ProcessGroupWrapper is used when TORCH_DISTRIBUTED_DEBUG=DETAIL and used to detect mismatched collectives. It performs this by serializing each collective call (Operation type, tensor data types, device types, shapes) into a tensor, performing an AllGather among all other process group wrappers, then detecting whether any of these tensors are mismatched.


**TODO**: The log message can also log the tensor data type, shapes, devices as well using data from deserialize method. Will file an issue

[ghstack-poisoned]

H-Huang committed:
Update on "[BE] Update ProcessGroupWrapper to add deserializer and improve logs"


Fixes #78022
And solves #72435 improvement #3

### Current Error Message:
`Detected mismatch between collectives on ranks. Rank 3 is running inconsistent collective: CollectiveFingerPrint(OpType=BARRIER)`

### New Error Message (shows rank and collective):
`Detected mismatch between collectives on ranks. Rank 0 is running collective: CollectiveFingerPrint(OpType=ALLREDUCE, …), but Rank 1 is running collective: REDUCE.`

### Changes:
- Add a deserialize function to deserialize a tensor to a CollectiveFingerprint
- Update log message to output the RANK and COLLECTIVE TYPE

### Context: 
ProcessGroupWrapper is used when TORCH_DISTRIBUTED_DEBUG=DETAIL and used to detect mismatched collectives. It performs this by serializing each collective call (Operation type, tensor data types, device types, shapes) into a tensor, performing an AllGather among all other process group wrappers, then detecting whether any of these tensors are mismatched.


**TODO**: The log message can also log the tensor data type, shapes, devices as well using data from deserialize method. Will file an issue

[ghstack-poisoned]

H-Huang committed:
Update on "[BE] Update ProcessGroupWrapper to add deserializer and improve logs"


Fixes #78022
And solves #72435 improvement #3

### Current Error Message:
`Detected mismatch between collectives on ranks. Rank 3 is running inconsistent collective: CollectiveFingerPrint(OpType=BARRIER)`

### New Error Message (shows rank and collective):
`Detected mismatch between collectives on ranks. Rank 0 is running collective: CollectiveFingerPrint(OpType=ALLREDUCE, …), but Rank 1 is running collective: REDUCE.`

### Changes:
- Add a deserialize function to deserialize a tensor to a CollectiveFingerprint
- Update log message to output the RANK and COLLECTIVE TYPE

### Context: 
ProcessGroupWrapper is used when TORCH_DISTRIBUTED_DEBUG=DETAIL and used to detect mismatched collectives. It performs this by serializing each collective call (Operation type, tensor data types, device types, shapes) into a tensor, performing an AllGather among all other process group wrappers, then detecting whether any of these tensors are mismatched.


**TODO**: The log message can also log the tensor data type, shapes, devices as well using data from deserialize method. Will file an issue

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge
